server/src/main/java/com/linecorp/centraldogma/server/internal/MetadataRolesMigrationPlugin.java

```diff
@@ -0,0 +1,127 @@
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Stopwatch;
+import com.google.common.collect.ImmutableMap;
+
+import com.linecorp.armeria.common.util.UnmodifiableFuture;
+import com.linecorp.centraldogma.common.Author;
+import com.linecorp.centraldogma.common.Change;
+import com.linecorp.centraldogma.common.Entry;
+import com.linecorp.centraldogma.common.Markup;
+import com.linecorp.centraldogma.common.Query;
+import com.linecorp.centraldogma.common.Revision;
+import com.linecorp.centraldogma.internal.Jackson;
+import com.linecorp.centraldogma.server.CentralDogmaConfig;
+import com.linecorp.centraldogma.server.command.Command;
+import com.linecorp.centraldogma.server.command.CommandExecutor;
+import com.linecorp.centraldogma.server.command.CommitResult;
+import com.linecorp.centraldogma.server.metadata.MetadataService;
+import com.linecorp.centraldogma.server.plugin.Plugin;
+import com.linecorp.centraldogma.server.plugin.PluginContext;
+import com.linecorp.centraldogma.server.plugin.PluginTarget;
+import com.linecorp.centraldogma.server.storage.project.InternalProjectInitializer;
+import com.linecorp.centraldogma.server.storage.project.Project;
+import com.linecorp.centraldogma.server.storage.repository.Repository;
+
+public final class MetadataRolesMigrationPlugin implements Plugin {
+
+    private static final Logger logger = LoggerFactory.getLogger(MetadataRolesMigrationPlugin.class);
+
+    @VisibleForTesting
+    static final String METADATA_ROLES_MIGRATION_JOB_LOG =
+            "/metadata-roles-job.json";
+
+    @Override
+    public PluginTarget target(CentralDogmaConfig config) {
+        return PluginTarget.LEADER_ONLY;
+    }
+
+    @Override
+    public CompletionStage<Void> start(PluginContext context) {
+        if (hasMigrationLog(context)) {
+            logger.debug("Metadata roles have already been migrated. Skipping auto migration...");
+            return UnmodifiableFuture.completedFuture(null);
+        }
+        logger.info("Starting metadata roles migration ...");
+
+        final MetadataService metadataService = new MetadataService(context.projectManager(),
+                                                                    context.commandExecutor());
+        final Stopwatch stopwatch = Stopwatch.createStarted();
+        for (Project project : context.projectManager().list().values()) {
+            final String name = project.name();
+            try {
+                metadataService.migrateMetadata(name).join();
+            } catch (Exception ex) {
+                // No need to rollback because the migration is not committed. Just log the error.
+                logger.warn("Failed to migrate metadata roles of {}", name, ex);
+                return UnmodifiableFuture.completedFuture(null);
```

Comment:
Question)

1. If a metadata is already migrated, will a `RedundantChangeException` be thrown?
2. If an exception is thrown, would it be a better idea to log and continue instead of returning? Whether the metadata was migrated or not doesn't seem to affect the behavior.

Reply:
> If a metadata is already migrated, will a RedundantChangeException be thrown?

Just the head revision is returned.
https://github.com/line/centraldogma/blob/a813dcd7de41670b2303681808be512897fa930a/server/src/main/java/com/linecorp/centraldogma/server/metadata/RepositorySupport.java#L128-L132

> If an exception is thrown, would it be a better idea to log and continue instead of returning? Whether the metadata was migrated or not doesn't seem to affect the behavior.

Yeah, we can probably do that. Let me make a change. 😉